### ISSUE TITLE
Wishlist: shortcut for combining map/reduce/filter/n-values with single iteration

### ISSUE THREAD

**Issue by elfprince13** [NetLogo/NetLogo#405]
It would be very useful to be able to chain the abilities of the task-oriented primitives mentioned in the Issue title. NetLogo lends itself to functional programming (by way of its Logo heritage and support for first-class-functions in tasks), but using common functional idioms can cause performance to suffer as compared to implementing the same algorithm less cleanly with repeat or while. Being able to chain the iteration of these primitives into a single loop (in the internal Java/Scala implementation) would reduce a lot of the associated overhead.

For example, `map` looks like this:

```
// make the result list.
val result = new LogoListBuilder
var i = 0
val actuals = new Array[AnyRef](n)
while(i < size) {
  var j = 0
  while(j < n) {
    actuals(j) = iters(j).next()
    j += 1
  }
  result.add(task.report(context, actuals))
  i += 1
}
result.toLogoList
```

and `reduce` looks like this:

```
val it = list.iterator
var result = it.next()
while (it.hasNext)
  result = task.report(context, Array(result, it.next()))
result
```

A combination of the two would look something like this:

```
// make the result list.
var result = nil
var i = 0
val actuals = new Array[AnyRef](n)
while(i < size) {
  var j = 0
  while(j < n) {
    actuals(j) = iters(j).next()
    j += 1
  }
  var mapped_val = maptask.report(context, actuals)
  result = if(result == nil){ mapped_val } else { reducetask.report(context, Array(result, mapped_val)) }
  i += 1
}
result
```

Many of the possible permutations of these primitives would be useful in some form or another (for example: I also commonly find myself doing things like `(map [position-dependent task] (n-values (length somelist) [?]) somelist)`). Barring a more-general solution, having both an `enumerate` primitive (`map` with an extra positional argument) and a `map-reduce` primitive would be exceedingly useful as probably the two most common such permutations.

**Comment by mrerrormessage** [comment]
related: #679

**Comment by ERSUCC** [comment]
If I'm understanding the requests correctly, this functionality is now available in the [fp](https://github.com/NetLogo/FP-Extension) extension.

**Comment by TheBizzle** [comment]
Nah.  What this issue is requesting is something closer to [collection views in Scala](https://www.baeldung.com/scala/stream-vs-views-vs-iterators#view).  In a lazy language (e.g. Haskell), you get this sort of composition of function performance for free.  In a strict language, you have to work harder for it.

It's kind of doubtful that we would add a language construct to NetLogo to fix this.  More likely, we could add some basic form of it with some compiler optimizations.  Or maybe an extension could be built to introduce a lazy list or "list view" data type. 🤷‍♂️
